services/httpd/msgp_emitter.go

```diff
@@ -0,0 +1,160 @@
+package httpd
+
+import (
+	"io"
+	"net/http"
+
+	"github.com/influxdata/influxdb/influxql"
+	"github.com/tinylib/msgp/msgp"
+)
+
+type messagePackEncoder struct {
+	Epoch     string
+	ChunkSize int
+}
+
+func (e *messagePackEncoder) ContentType() string {
+	return "application/x-msgpack"
+}
+
+func (e *messagePackEncoder) Encode(w io.Writer, header ResponseHeader, results <-chan *influxql.ResultSet) {
+	var convertToEpoch func(row *influxql.Row)
+	if e.Epoch != "" {
+		convertToEpoch = epochConverter(e.Epoch)
+	}
+	values := make([][]interface{}, 0, e.ChunkSize)
+
+	enc := msgp.NewWriter(w)
+	enc.WriteMapHeader(1)
+	enc.WriteString("results")
+	enc.WriteInt(header.Results)
+
+	enc.Flush()
+	if w, ok := w.(http.Flusher); ok {
+		w.Flush()
```

Comment:
It's possible to know the response header. I don't know about other HTTP libraries, but the Go one will block until it receives the headers. Once it receives the headers, you can read the body whenever you want.

So it's mostly to make Ctrl+C functionality possible. If you don't flush the headers immediately, the client won't be able to get to the section of code where handling an interrupt is possible.

A (non-compiling) example:

```
resp, err := http.DefaultClient.Do(req)
if err != nil {
    panic(err)
}
// Does not get here until the headers are sent back.
signals := make(os.Signal, 0, 4)
signal.Notify(signals, os.Interrupt)
go func() {
    <-signals
    resp.Body.Close()
}()

for {
    // read points from the stream
}
```

Something like that, but the code above doesn't work and has issues.